# Lecture 3

## Exercise 3.1

Write a function that takes two lists of strings and returns a list of those strings that are present in both input lists.

In [97]:
list1 = ["a", "b", "c", "d", "e"]
list2 = ["a", "f", "g", "d", "h"]

def list_present(list1,list2):
    return [i for i in list1 if i in list2]
    
print(f'(Exercise 3.1) {list_present(list1,list2)}')

(Exercise 3.1) ['a', 'd']


## Exercise 3.2

Write a function that takes a list of numbers as an argument and returns the mean of the numbers.

In [98]:
list1 = [5, 3, 7, 8, 6]

def list_ave(list1):
    return sum(list1)/len(list1)
    
print(f'(Exercise 3.2) {list_ave(list1)}')

(Exercise 3.2) 5.8


## Exercise 3.3

Write a script that contains a function, the function should:

>• Accept a name of the day of the week and return the next day of the week (e.g., given “Monday”, return “Tuesday”), hint: use mylist.index(myelement)

>• How would you handle the possibility that the argument to the function is misspelled?

>• Make sure that if the user inputs “Sunday”, "Monday" is returned as the next day.
Assume that the week begins with a Monday and ends with a Sunday.


In [99]:
days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

def input_day(day):
    return days[(days.index(day) + 1) % len(days)] if day in days else 'Typo, please try again!'
print(f"(Exercise 3.3) {input_day('Tuesday')}")

(Exercise 3.3) Wednesday


## Exercise 3.4

Write a function that takes a list of numbers and returns the variance of the numbers (reuse the previous function from Exercise 3.2, if possible). The formula for variance is:

In [102]:
list1 = [5, 3, 7, 8, 6]

def list_var(list1):
    return [round(((x-list_ave(list1))/len(list1)),2) for x in list1]
list_var(list1)

[-0.16, -0.56, 0.24, 0.44, 0.04]

## Exercise 3.5

Consider the class Card below. Your task is to use this class to create a deck of cards: create an empty list called "deck", and then iterate over all the ranks and suits to create all the necessary cards and add them all to the deck.

In [112]:
import random

class Card:
    
    def __init__(self, suit, rank):
        self.suit = suit
        self.rank = rank

suits = ["Spades", "Clubs", "Hearts", "Diamonds"]
ranks = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"]

deck = [Card(suit,rank) for suit in suits for rank in ranks]

After that, you need to shuffle the deck, take the first card and print its suit and rank. The format of the output should be as follows: "The Q of Hearts".
To shuffle the deck of cards, you can use the built-in "random" package as follows:

In [113]:
random.shuffle(deck)

print(f'The {deck[0].rank} of {deck[0].suit}')

The A of Hearts


## Exercise 3.6

Write a program that reads the contents of "aston.txt" (see the attachment), and counts occurrences of every word in the text, and prints the 10 most common words.

In [100]:
with open('Aston (Lecture 3).txt','r') as reader:
    text = reader.read().split(' ')
    
def text_counter(text):
    return [[len(x),x] for x in text]
sorted(text_counter(text),reverse=True)[:10]

[[29, 'UK.[10]\n\nHistory\n\nPredecessor'],
 [25, 'buildings.[14][unreliable'],
 [21, '1965.[16]\n\nUniversity'],
 [18, 'graduates.[22][23]'],
 [17, 'institutions\n\nThe'],
 [17, 'University\n\nAston'],
 [17, 'Sciences.[21]\n\nIn'],
 [16, 'University.\n\nThe'],
 [16, 'Forward.[17]\n\nIn'],
 [15, 'technology.\n\nIn']]

## Exercise 3.7

Find out the number of days between 10th July 2018 and 20th September 2019

In [32]:
from datetime import datetime
(datetime.strptime('20-9-2019','%d-%m-%Y') - datetime.strptime('10-7-2018','%d-%m-%Y')).days

437

## Exercise 3.8

In the period between 1 January 2018 and 29 September 2019, an investment into stock A produced a cumulative return of 0.29, i.e. the investment grew by 29%. In the period between 1 February 2018 and 31 July 2019, an investment into stock B produced a cumulative return of 0.23. Calculate the annualized returns for both stocks and find out which stock proved the better investment.

The formula for annualized return is:

AnnualizedReturn = (1 + CumulativeReturn)365/DaysHeld - 1

where DaysHeld is the number of days the investment was held in the stock.

In [37]:
CumulativeReturn1 = 0.29
CumulativeReturn2 = 0.23
DaysHeld1 = (datetime.strptime('20-9-2019','%d-%m-%Y') - datetime.strptime('10-7-2018','%d-%m-%Y')).days
DaysHeld2 = (datetime.strptime('31-7-2019','%d-%m-%Y') - datetime.strptime('1-2-2018','%d-%m-%Y')).days
AnnualizedReturn1 = round((1 + CumulativeReturn1)**(365/DaysHeld1) - 1,2)
AnnualizedReturn2 = round((1 + CumulativeReturn2)**(365/DaysHeld2) - 1,2)

print(f'Stock A ({AnnualizedReturn1}%) is the better investment than Stock B ({AnnualizedReturn2}%)') if AnnualizedReturn1 > AnnualizedReturn2 else print(f'Stock B ({AnnualizedReturn2}%) is the better investment than Stock A ({AnnualizedReturn1}%)') 

Stock A (0.24%) is the better investment than Stock B (0.15%)


## Exercise 3.9

Use the "factorial" function of the built-in "math" package to calculate the factorial of 5. The function takes a single argument, the number for which the factorial needs to be calculated.

Import the function in at least two different ways.

In [30]:
import math

factorial_ = math.factorial(5)
print(factorial_)

120


In [31]:
from math import factorial

factorial_ = factorial(5)
print(factorial_)

120


## Exercise 3.10

The "requests" package contains a function called "get". The function accepts a URL as a string argument and returns a data structure with different types of information on the URL.
AnnualizedReturn
Import the function and retrieve a URL (e.g., "https://www.aston.ac.uk"), use the variable named "response" to store the data structure and then save the value "response.text" to a local file called "url.html".

In [29]:
import requests
from bs4 import BeautifulSoup

URL = "https://www.aston.ac.uk"
response = requests.get(URL)
soup = BeautifulSoup(response.content, "html.parser")

with open('url.html','x',encoding='utf-8') as creator:
    with open('response.txt','w') as writer:
        writer.write(soup.prettify())